### PR TITLE
Added upper bound for maximal compression rate of ACTL

### PR DESCRIPTION
**Pull request by nikita-savelyevv**
- Added `maximal_compression_rate` property for compression controllers. It returns the maximal compression rate that controller can compress the model.
- Added an upper bound for maximal compression rate Adaptive Compression Training Loop may try to reach
- Added some related tests

### Reason for changes

During ACTL, the model compression rate may get arbitrary high. But not all of the compression controllers support that, e.g. pruning controller with FLOPS target. This lead to a runtime error.

### Related tickets

98894

### Tests

Added the following tests:
- a test for correctness of computation of `compression_controller.maximal_compression_rate` (PT + TF)
- a test for ACTL where maximal compression rate is reached and algorithm still finished correctly

### CODE REVIEW

**Comment by nikita-savelyevv** [comment]
run openvino pre-commit tests

File: nncf/api/compression.py [hunk 1]
```diff
@@ -315,6 +315,13 @@ def disable_scheduler(self) -> None:
         it to a dummy one that does not change the compression rate.
         """
 
+    @property
+    @abstractmethod
+    def maximal_compression_rate(self) -> float:
+        """
+        Returns the maximal model compression rate supported by the compression controller
```
Comment (daniil-lyakhov) [inline nncf/api/compression.py:322]:
```suggestion
        Returns the maximal model compression rate supported by the compression controller.
```

File: nncf/tensorflow/pruning/filter_pruning/algorithm.py [hunk 2]
```diff
@@ -240,6 +240,12 @@ def set_pruning_level(self, pruning_level: float,
         if run_batchnorm_adaptation:
             self._run_batchnorm_adaptation()
 
+    @property
+    def maximal_compression_rate(self) -> float:
+        max_flops, _ = self._calculate_flops_and_weights_in_uniformly_pruned_model(1.0)
+        max_compression_rate = 1 - max_flops / max(self.full_flops, 1)
+        return max_compression_rate
```
Comment (daniil-lyakhov) [inline nncf/tensorflow/pruning/filter_pruning/algorithm.py:247]: 1) `max_flops` is a bit confusing name because it is minimum possible flops for the model after the pruning. `min_flops` is clearer in my opinion
2) `max(self.full_flops, 1)` is needed for a case when a model has 0 flops, right?

**Comment by daniil-lyakhov** [comment]
By the way: according to `get_rounded_pruned_element_number` in case sparsity rate is 1. it always returns `total` elements even if it's not a multiply of 8. Is it bug or feature? @alexsu52

File: nncf/tensorflow/pruning/filter_pruning/algorithm.py (hunk 2 above)
Comment (nikita-savelyevv) [reply nncf/tensorflow/pruning/filter_pruning/algorithm.py:247]: 1. I see. If I change, it will be inconsistent with this part [here](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/pruning/filter_pruning/algorithm.py#L143) . Should it be changed it to `min_` in both places?
2. Yes

Comment (daniil-lyakhov) [reply nncf/tensorflow/pruning/filter_pruning/algorithm.py:247]: 1) Yes
2) 👍

Comment (nikita-savelyevv) [reply nncf/tensorflow/pruning/filter_pruning/algorithm.py]: Following that trail, decided to put maximal compression rate computation inside existing `PrunedModelTheoreticalBorderline` where it seems to belong

File: nncf/tensorflow/pruning/filter_pruning/algorithm.py [hunk 3]
```diff
@@ -240,6 +240,13 @@ def set_pruning_level(self, pruning_level: float,
         if run_batchnorm_adaptation:
             self._run_batchnorm_adaptation()
 
+    @property
+    def maximal_compression_rate(self) -> float:
+        stats = PrunedModelTheoreticalBorderline(
+            self._pruned_layers_num, self._prunable_layers_num, self._min_prunable_flops,
+            self._min_prunable_params, self.full_flops, self.full_params_num)
+        return stats.maximal_pruning_rate
```
Comment (daniil-lyakhov) [inline nncf/tensorflow/pruning/filter_pruning/algorithm.py:248]: I suggest to calculate `maximal_pruning_rage` explicitly. Copy paste will be unified in future when common pruning algorithm builder/controller will be introduced.

Comment (nikita-savelyevv) [reply nncf/tensorflow/pruning/filter_pruning/algorithm.py]: I agree, this is more clean

File: nncf/torch/pruning/filter_pruning/algo.py [hunk 4]
```diff
@@ -613,6 +613,11 @@ def disable_scheduler(self):
         self._scheduler = StubCompressionScheduler()
         self._scheduler.current_pruning_level = 0.0
 
+    @property
+    def maximal_compression_rate(self) -> float:
+        max_compression_rate = 1 - self._min_prunable_flops / max(self.full_flops, 1) if self.prune_flops else 1.0
+        return max_compression_rate
```
Comment (daniil-lyakhov) [inline nncf/torch/pruning/filter_pruning/algo.py:619]: IMHO
```suggestion
        if self.prune_flops:
            return 1.0
        return 1 - self._min_prunable_flops / max(self.full_flops, 1)
```

**Review by daniil-lyakhov — APPROVED** [review]

File: nncf/torch/pruning/filter_pruning/algo.py (hunk 4 above)
Comment (nikita-savelyevv) [reply nncf/torch/pruning/filter_pruning/algo.py]: 👍

**Review by daniil-lyakhov — APPROVED** [review]

**Review by ljaljushkin — CHANGES_REQUESTED** [review]
Why not make it more consistent and implement `maximal_compression_rate` for all other algorithms that have overridden `compression_rate` method?
I think this method is needed for `magnitude`, `rb` and `movement` sparsity algorithms.

**Comment by ljaljushkin** [comment]
> Why not make it more consistent and implement `maximal_compression_rate` for all other algorithms that have overridden `compression_rate` method? I think this method is needed for `magnitude`, `rb` and `movement` sparsity algorithms.

maximal_sparsity_rate = 1 should be ok for sparsity algorithms.

**Review by ljaljushkin — APPROVED** [review]

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 5]
```diff
@@ -209,6 +209,11 @@ def __init__(self,
             raise RuntimeError('No compression algorithm supported by the accuracy-aware training '
                                'runner was specified in the config')
 
+        if maximal_compression_rate > self.adaptive_controller.maximal_compression_rate:
+            maximal_compression_rate = self.adaptive_controller.maximal_compression_rate
+            nncf_logger.warning(f'Updated training loop maximal compression rate by the maximal compression rate '
```
Comment (alexsu52) [inline nncf/common/accuracy_aware_training/training_loop.py:214]: Probably, the maximum compression rate in 95% can not be archived for most models. I mean that the user will see warning if maximal_compression_rate will not defined and use default value 0.95. It can be confuse an user. I would recommend to handle it during the accuracy aware loop.

File: nncf/common/pruning/statistics.py [hunk 6]
```diff
@@ -161,26 +161,26 @@ class PrunedModelTheoreticalBorderline(Statistics):
     def __init__(self,
                  num_pruned_layers: int,
                  num_prunable_layers: int,
-                 max_prunable_flops: float,
-                 max_prunable_params: float,
+                 min_prunable_flops: float,
+                 min_prunable_params: float,
```
Comment (alexsu52) [inline nncf/common/pruning/statistics.py:165]: Seems like `max_prunable_flops` makes more sense. 

But you can rename them as following, for example:
```suggestion
                 min_flops: float,
                 min_params: float,
```

File: nncf/tensorflow/pruning/filter_pruning/algorithm.py [hunk 7]
```diff
@@ -240,6 +240,12 @@ def set_pruning_level(self, pruning_level: float,
         if run_batchnorm_adaptation:
             self._run_batchnorm_adaptation()
 
+    @property
+    def maximal_compression_rate(self) -> float:
+        if self.prune_flops:
+            return 1 - self._min_prunable_flops / max(self.full_flops, 1)
+        return 1.0
```
Comment (alexsu52) [inline nncf/tensorflow/pruning/filter_pruning/algorithm.py:247]: This is not correct statement. The pruning algorithm can not prune all model parameters.

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: nncf/api/compression.py [hunk 8]
```diff
@@ -315,6 +315,13 @@ def disable_scheduler(self) -> None:
         it to a dummy one that does not change the compression rate.
         """
 
+    @property
+    @abstractmethod
+    def maximal_compression_rate(self) -> float:
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/api/compression.py:320]: Since accuracy aware algorithm uses common compression interface, we can't introduce it for specific algorithms and not introduce for others.
`maximal_compression_rate`=1 is valid for sparsity, it means there's no technical limitation to compress everything.

The only concern is that nobody needs maximal compression rate equal to 1. This model will not make any sense.
Probably we should call it `upper_bound_for_compression_rate` or something like that.

@nikita-savelyevv @alexsu52 What do you think about another approach when there's no `maximal_compression_rate`, but everything is handled by set_compression_rate?
We can return a specific error or exception that means we set more than the algorithm can really compress.

File: nncf/tensorflow/pruning/filter_pruning/algorithm.py (hunk 7 above)
Comment (nikita-savelyevv) [reply nncf/tensorflow/pruning/filter_pruning/algorithm.py:247]: Yes. But at the same time currently it is possible to set any (channel) pruning rate and it will not throw an exception to it. I think this logic needs to be implemented first in order to support the behavior of the property value you suggest.

File: nncf/api/compression.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/api/compression.py:320]: @daniil-lyakhov @alexsu52, should `maximal_compression_rate` be different when some layers are in ignored scope for filter pruning by channels and for sparsity?

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 5 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:214]: Agree. Let's keep only the logging during the loop itself.

File: nncf/common/pruning/statistics.py (hunk 6 above)
Comment (nikita-savelyevv) [reply nncf/common/pruning/statistics.py:165]: `max_prunable_flops` is not quite correct, since this value represents the minimal FLOPS we are able to reach with the maximum compression rate. I don't have problems with renaming it to `min_flops` though.

File: nncf/api/compression.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/api/compression.py:320]: In an offline discussion, it was decided to keep the introduction of `maximal_compression_rate` property as it may be useful in the future. The property should also be implemented for the magnitude sparsity compression algorithm, because for it, the property is not equal to 1.0 by default.

File: nncf/tensorflow/sparsity/magnitude/algorithm.py [hunk 9]
```diff
@@ -190,7 +190,7 @@ def _select_threshold(self, sparsity_level):
         if not all_weights:
             return 0.0
         all_weights_tensor = tf.sort(tf.concat(all_weights, 0))
-        index = int(tf.cast(tf.size(all_weights_tensor), all_weights_tensor.dtype) * sparsity_level)
+        index = int(tf.cast(tf.size(all_weights_tensor) - 1, all_weights_tensor.dtype) * sparsity_level)
```
Comment (nikita-savelyevv) [inline nncf/tensorflow/sparsity/magnitude/algorithm.py:193]: Fixed a minor bug. Not consistent with PT [implementation ](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/magnitude/algo.py#L145)

File: nncf/common/pruning/statistics.py (hunk 6 above)
Comment (alexsu52) [reply nncf/common/pruning/statistics.py:165]: Possible, it depends on an interpretation. My opinion is the max prunable flops is number of flops when prunable flops were pruned to their maximum.

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 10]
```diff
@@ -209,6 +209,9 @@ def __init__(self,
             raise RuntimeError('No compression algorithm supported by the accuracy-aware training '
                                'runner was specified in the config')
 
+        if maximal_compression_rate > self.adaptive_controller.maximal_compression_rate:
+            maximal_compression_rate = self.adaptive_controller.maximal_compression_rate
+
```
Comment (alexsu52) [inline nncf/common/accuracy_aware_training/training_loop.py:214]:
```suggestion
        maximal_compression_rate = min(
            maximal_compression_rate, 
            self.adaptive_controller.maximal_compression_rate)
```

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 11]
```diff
@@ -209,6 +209,9 @@ def __init__(self,
             raise RuntimeError('No compression algorithm supported by the accuracy-aware training '
                                'runner was specified in the config')
 
+        if maximal_compression_rate > self.adaptive_controller.maximal_compression_rate:
+            maximal_compression_rate = self.adaptive_controller.maximal_compression_rate
```
Comment (alexsu52) [inline nncf/common/accuracy_aware_training/training_loop.py:213]: Just to clarify, AdaptiveCompressiontrainingLoop is ablle to reach the `maximal_compression_rate `?

File: nncf/tensorflow/sparsity/magnitude/algorithm.py [hunk 12]
```diff
@@ -256,6 +256,31 @@ def compression_stage(self) -> CompressionStage:
             return CompressionStage.FULLY_COMPRESSED
         return CompressionStage.PARTIALLY_COMPRESSED
 
+    @property
+    def maximal_compression_rate(self) -> float:
```
Comment (alexsu52) [inline nncf/tensorflow/sparsity/magnitude/algorithm.py:260]: Could you explain why this implementation is consistent with "compression_rate"?

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 11 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:213]: If model compresses successfully, it will reach `(maximal_compression_rate - compression_rate_step)` compression rate, where `compression_rate_step` is the last compression rate change step of the algo. For example, if the compression rate is only increasing then the algorithm will exit at `(maximal_compression_rate - 0.1)` compression rate by default.

File: nncf/common/pruning/statistics.py (hunk 6 above)
Comment (nikita-savelyevv) [reply nncf/common/pruning/statistics.py]: Renamed to `min_possible_*`. It is a good candidate and the variables are already used with that name inside `PrunedModelTheoreticalBorderline`.

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 10 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py]: Done

File: nncf/tensorflow/sparsity/magnitude/algorithm.py (hunk 12 above)
Comment (nikita-savelyevv) [reply nncf/tensorflow/sparsity/magnitude/algorithm.py]: Decided to return `maximal_compression_rate` of 1.0 for sparsity compression algorithm
